Failed build ```
FAILED: fdbrpc/CMakeFiles/fdbrpc.dir/HTTP.cpp.o 
/usr/local/bin/ccache /usr/local/bin/clang++ -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTEXT_NO_LIB -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_NO_LIB -DBOOST_SERIALIZATION_NO_LIB -DBOOST_SYSTEM_NO_DEPRECATED -DBOOST_SYSTEM_NO_LIB -DBOOST_URL_NO_LIB -DCARES_STATICLIB -DFLOW_GRPC_ENABLED -DNO_INTELLISENSE -DCOMPILATION_UNIT=/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/build_output/fdbrpc/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/libeio -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/flow/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/build_output/flow/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/contrib/libb64/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/build_output/_deps/fmt-src/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/contrib/SimpleOpt/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/contrib/crc32/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/contrib/md5/include -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/contrib/rapidjson -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/build_output/generated -I/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/libcoroutine -isystem /opt/boost_1_86_0_clang/include -isystem /opt/grpc_clang/include -stdlib=libc++ -O3 -DNDEBUG -std=c++20 -DCMAKE_BUILD -fno-omit-frame-pointer -gdwarf-4 -ggdb1 -gz -mavx -Wall -Wextra -Wredundant-move -Wpessimizing-move -Woverloaded-virtual -Wshift-sign-overflow -Wno-sign-compare -Wno-undefined-var-template -Wno-unknown-warning-option -Wno-unused-parameter -Wno-constant-logical-operand -Wno-deprecated-copy -Wno-delete-non-abstract-non-virtual-dtor -Wno-range-loop-construct -Wno-reorder-ctor -Wno-unused-command-line-argument -Wno-ambiguous-reversed-operator -Wno-vla-cxx-extension -Wno-unused-function -Wno-unused-private-field -Wno-nullability-completeness -Wno-macro-redefined -Wno-register -Werror -Wno-error=format -Wunused-variable -Wno-deprecated -fvisibility=hidden -Wreturn-type -fPIC -DHAVE_OPENSSL -MD -MT fdbrpc/CMakeFiles/fdbrpc.dir/HTTP.cpp.o -MF fdbrpc/CMakeFiles/fdbrpc.dir/HTTP.cpp.o.d -o fdbrpc/CMakeFiles/fdbrpc.dir/HTTP.cpp.o -c /codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp
distcc[2770] ERROR: compile /codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp on distcc-rhel9-fd4fc40438b568fb.elb.us-west-2.amazonaws.com/96 failed
distcc[2770] (dcc_build_somewhere) Warning: remote compilation of '/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp' failed, retrying locally
distcc[2770] Warning: failed to distribute /codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp to distcc-rhel9-fd4fc40438b568fb.elb.us-west-2.amazonaws.com/96, running locally instead
distcc[2770] ERROR: compile /codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp on localhost failed
/codebuild/output/src3520963078/src/github.com/apple/foundationdb/fdbrpc/HTTP.cpp:241:73: error: use of undeclared identifier 'buggify'
  241 |                 if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && buggify(0.01)) {
      |                                                                                       ^
1 error generated.
```

--- a/fdbrpc/HTTP.cpp
+++ b/fdbrpc/HTTP.cpp
@@ -24,9 +24,10 @@
 #include "flow/IRandom.h"
 #include "flow/Net2Packet.h"
 #include "flow/Trace.h"
+#include "flow/Knobs.h"
+#include "flow/CodeProbe.h"
 #include "md5/md5.h"
 #include "libb64/encode.h"
-#include "flow/Knobs.h"
 #include <cctype>
 #include <sstream>
 #include "flow/IConnection.h"
@@ -227,7 +228,7 @@ Future<Void> writeResponse(Reference<IConnection> conn, Reference<OutgoingRespon
 	}
 }
 
-// Read at least 1 bytes from conn and up to maxlen in a single read, append read data into *buf
+// Read at least 1 byte from conn and up to maxlen in a single read, append read data into *buf
 // Returns the number of bytes read.
 Future<int> read_into_string(Reference<IConnection> conn, std::string* buf, int maxlen) {
 	while (true) {
@@ -235,7 +236,12 @@ Future<int> read_into_string(Reference<IConnection> conn, std::string* buf, int 
 		int originalSize = buf->size();
 		// TODO:  resize is zero-initializing the space we're about to overwrite, so do something else, which probably
 		// means not using a string for this buffer
-		// FIXME: buggify read size as well
+
+		// Buggify read size to exercise partial-read code paths in simulation
+		if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && BUGGIFY_WITH_PROB(0.01)) {
+			maxlen = deterministicRandom()->randomInt(1, 10);
+			CODE_PROBE(true, "HTTP buggified read size");
+		}
 		buf->resize(originalSize + maxlen);
 		uint8_t* wptr = (uint8_t*)buf->data() + originalSize;
 		int len = conn->read(wptr, wptr + maxlen);

--- a/fdbrpc/HTTP.cpp
+++ b/fdbrpc/HTTP.cpp
@@ -238,7 +238,7 @@ Future<int> read_into_string(Reference<IConnection> conn, std::string* buf, int 
 		// means not using a string for this buffer
 
 		// Buggify read size to exercise partial-read code paths in simulation
-		if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && BUGGIFY_WITH_PROB(0.01)) {
+		if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && buggify(0.01)) {
 			maxlen = deterministicRandom()->randomInt(1, 10);
 			CODE_PROBE(true, "HTTP buggified read size");
 		}

--- a/fdbrpc/HTTP.cpp
+++ b/fdbrpc/HTTP.cpp
@@ -214,7 +214,7 @@ Future<Void> writeResponse(Reference<IConnection> conn, Reference<OutgoingRespon
 	response->data.content->prependWriteBuffer(pFirst, pLast);
 	while (true) {
 		int trySend = FLOW_KNOBS->HTTP_SEND_SIZE;
-		if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && buggify(0.01)) {
+		if (g_network->isSimulated() && !g_simulator->speedUpSimulation && buggify(0.1)) {
 			trySend = deterministicRandom()->randomInt(1, 10);
 		}
 		int len = conn->write(response->data.content->getUnsent(), trySend);
@@ -238,8 +238,8 @@ Future<int> read_into_string(Reference<IConnection> conn, std::string* buf, int 
 		// means not using a string for this buffer
 
 		// Buggify read size to exercise partial-read code paths in simulation
-		if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && buggify(0.01)) {
-			maxlen = deterministicRandom()->randomInt(1, 10);
+		if (g_network->isSimulated() && !g_simulator->speedUpSimulation && maxlen > 1 && buggify(0.1)) {
+			maxlen = deterministicRandom()->randomInt(1, maxlen);
 			CODE_PROBE(true, "HTTP buggified read size");
 		}
 		buf->resize(originalSize + maxlen);
@@ -666,7 +666,7 @@ Future<Reference<HTTP::IncomingResponse>> doRequestActor(Reference<IConnection> 
 			}
 
 			int trySend = FLOW_KNOBS->HTTP_SEND_SIZE;
-			if ((!g_network->isSimulated() || !g_simulator->speedUpSimulation) && buggify(0.01)) {
+			if (g_network->isSimulated() && !g_simulator->speedUpSimulation && buggify(0.1)) {
 				trySend = deterministicRandom()->randomInt(1, 10);
 			}
 			co_await sendRate->getAllowance(trySend);

